### PR TITLE
nodejs-week3

### PR DESCRIPTION
**Pull request by Sana-Shabeel**

### CODE REVIEW

File: src/backend/api/meals.js [hunk 1]
```diff
@@ -2,14 +2,203 @@ const express = require("express");
...
+      }
+    }
+    return validData;
   } catch (error) {
-    throw error;
+    console.log(error);
   }
+}
+
+// GET route all meals route
```
Comment (miltonhultgren) [inline src/backend/api/meals.js]: I would love to see the route definition and route handler separated.

```js
router.get("/", listMeals);

async function listMeals(req, res) => {
```

File: src/backend/api/meals.js [hunk 2]
```diff
@@ -2,14 +2,203 @@ const express = require("express");
...
+    // looping through the data object
+    for (const key in data) {
+      // checking if the mealsColumn has the key
+      if (mealsColumns.has(key)) {
+        validData[key] = data[key];
+      }
+    }
+    if ("price" in validData) {
+      if (typeof validData.price !== "number") {
+        throw new Error();
```
Comment (miltonhultgren) [inline src/backend/api/meals.js:27]: Make sure you always include a message in your errors, so you know what went wrong.

File: src/backend/api/meals.js [hunk 3]
```diff
@@ -2,14 +2,203 @@ const express = require("express");
...
+    }
+    return validData;
   } catch (error) {
-    throw error;
+    console.log(error);
   }
+}
+
+// GET route all meals route
+router.get("/", async (req, res) => {
```
Comment (miltonhultgren) [inline src/backend/api/meals.js:46]: This route should be able to handle a combination of any supported parameters.
Your code currently only support a subset mix and the "if based" code is fairly hard to follow.

How can you change this code so that it doesn't matter which parameters are passed in?
A combination of default parameter values and a combined knex query should be able to handle this in much less code!

File: src/backend/api/meals.js [hunk 4]
```diff
@@ -2,14 +2,203 @@ const express = require("express");
...
+        "location",
+        "when",
+        "created",
+        "price",
+        "max_reservation",
+        knex.raw("SUM(number_of_guests) AS sum")
+      )
+      .where("max_reservation", ">", "number_of_guests")
+      .groupBy("meal_id");
+    const availableMeals = getMeals.filter(
```
Comment (miltonhultgren) [inline src/backend/api/meals.js]: I'm pretty sure you can get knex to perform this filter for you!

File: src/backend/api/meals.js [hunk 5]
```diff
@@ -2,14 +2,203 @@ const express = require("express");
...
+    return res.json(limit);
+  }
+
+  res.json(meals);
 });
 
+// CREATE meal route
+router.post("/", async (req, res) => {
+  const validData = isValidData(req.body);
+  if (!validData) {
```
Comment (miltonhultgren) [inline src/backend/api/meals.js:88]: Perhaps isValidData should return two values instead:
`const [isValid, errors] = isValidData(req.body)`

As you perform each check, if one fails you add a string to the errors list and return the status at the end.

File: src/backend/api/meals.js [hunk 6]
```diff
@@ -2,14 +2,203 @@ const express = require("express");
...
+    });
+  }
+  res.json(meal);
+});
+
+// UPDATE meal route
+router.put("/:id", async (req, res) => {
+  const validData = isValidData(req.body);
+  // Check if id is included
+  const checkedId = await knex("meal").where({ id: req.params.id });
```
Comment (miltonhultgren) [inline src/backend/api/meals.js:123]: I'd do the validity check before touching the database,

**Review by miltonhultgren — CHANGES_REQUESTED** [review]
Mostly this looks good, there are a few small things you could improve and one big one.
The big one is that your endpoints should be able to handle any combination of parameters.

Further, I'd put some thought into what actions in your code are expensive (hint: asking the database!). So make sure you do the cheap validation (input data) before you ask the database anything.

File: src/backend/api/meals.js (hunk 3 above)
Comment (miltonhultgren) [reply src/backend/api/meals.js:46]: Also, review this comment about the usage of `if` https://github.com/negaren/hyf-homework/pull/19/files#r826736613

Comment (Sana-Shabeel) [reply src/backend/api/meals.js:46]: Are you saying there is a better way of handling the request.query without having to create an if statement for each and everyone of them ?

Comment (miltonhultgren) [reply src/backend/api/meals.js:46]: See https://github.com/negaren/meal-sharing/pull/1#discussion_r826793745 for more details.

Comment (miltonhultgren) [reply src/backend/api/meals.js:46]: Yes.
By moving the logic that converts and validates your parameters (and using default values) you can simplify this handler by a lot. 
See the comments I've linked above for more details on how.
